core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -286,19 +281,22 @@ private object RTS {
             val f: IO[Nothing, Option[List[Throwable]]] = f0.finalizer.run[Nothing, Unit].map(collectDefect)
             if (finalizer eq null) finalizer = f
             else
-              finalizer = finalizer.zipWith(f) {
-                case (Some(ts1), Some(ts2)) => Some(ts1 ++ ts2)
-                case (Some(ts1), None)      => Some(ts1)
-                case (None, Some(ts2))      => Some(ts2)
-                case (None, None)           => None
-              }
+              finalizer = finalizer.zipWith(f)(zipFailures)
           case _ =>
         }
       }
 
       finalizer
     }
 
+    private final def zipFailures(ts1: Option[List[Throwable]], ts2: Option[List[Throwable]]): Option[List[Throwable]] =
+      (ts1, ts2) match {
+        case (Some(ts1), Some(ts2)) => Some(ts1 ++ ts2)
+        case (ts @ Some(_), None)   => ts
+        case (None, ts @ Some(_))   => ts
+        case (None, None)           => None
+      }
+
```

Comment:
This is a re-implementation of one of the `Semigroup` instances for `Option[A]`.

Reply:
It's true, but now Scalaz depends on ZIO (which means, at the very least, no more impure code in Scalaz), so the reimplementation is necessary.